reactor-core/src/main/java/reactor/core/publisher/FluxPrefetch.java

```diff
@@ -0,0 +1,1404 @@
+			if (key == Attr.PREFETCH) return prefetch;
+			if (key == Attr.REQUESTED_FROM_DOWNSTREAM) return requested;
+			if (key == Attr.BUFFERED) return queue != null ? queue.size() : 0;
+			if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
+
+			return InnerOperator.super.scanUnsafe(key);
+		}
+
+		@Override
+		public CoreSubscriber<? super T> actual() {
+			return actual;
+		}
+
+		@Override
+		public void clear() {
+			if (sourceMode == Fuseable.ASYNC) {
+				queue.clear();
+				return;
+			}
+
+			// use guard on the queue instance as the best way to ensure there is no racing on draining
+			// the call to this method must be done only during the ASYNC fusion so all the callers will be waiting
+			// this should not be performance costly with the assumption the cancel is rare operation
+			if (DISCARD_GUARD.getAndIncrement(this) != 0) {
+				return;
+			}
+
+			int missed = 1;
+
+			for (; ; ) {
+				Operators.onDiscardQueueWithClear(queue, actual.currentContext(), null);
+
+				int dg = discardGuard;
+				if (missed == dg) {
+					missed = DISCARD_GUARD.addAndGet(this, -missed);
+					if (missed == 0) {
+						break;
+					}
+				}
+				else {
+					missed = dg;
+				}
+			}
+		}
+
+		@Override
+		public boolean isEmpty() {
+			return queue.isEmpty();
+		}
+
+		@Override
+		@Nullable
+		public T poll() {
+			if (sourceMode == Fuseable.NONE) {
+				if (firstRequest && prefetchMode == PrefetchMode.LAZY) {
+					firstRequest = false;
+					s.request(Operators.unboundedOrPrefetch(this.prefetch));
+				}
+
+				T value = queue.poll();
+				if (value != null) {
+					long p = produced + 1;
+					if (p == limit) {
+						produced = 0;
+						s.request(p);
+					}
+					else {
+						produced = p;
+					}
+				}
+				return value;
+			}
+			else {
+				return queue.poll();
+			}
+		}
+
+		public int requestFusion(int requestedMode) {
+			if (s instanceof QueueSubscription) {
+				@SuppressWarnings("unchecked") QueueSubscription<T> fusion = (QueueSubscription<T>) s;
+				int fusionMode = fusion.requestFusion(requestedMode);
+
+				if (fusionMode == Fuseable.SYNC) {
+					sourceMode = fusionMode;
+					outputMode = fusionMode;
+					queue = fusion;
+					done = true;
+					firstRequest = false;
+
+					WIP.lazySet(this, 0);
+					return fusionMode;
+				}
+				else if (fusionMode == Fuseable.ASYNC) {
+					sourceMode = fusionMode;
+					outputMode = fusionMode;
+					queue = fusion;
+					firstRequest = false;
+
+					WIP.lazySet(this, 0);
+					return fusionMode;
+				}
+			}
+
+			int fusionMode;
+			sourceMode = Fuseable.NONE;
+			queue = queueSupplier.get();
+
+			if ((requestedMode & Fuseable.ASYNC) != 0) {
+				outputMode = Fuseable.ASYNC;
+				fusionMode = Fuseable.ASYNC;
+			}
+			else {
+				fusionMode = Fuseable.NONE;
+			}
+
+			WIP.lazySet(this, 0);
+			return fusionMode;
+		}
+
+		@Override
+		public int size() {
+			return queue.size();
+		}
+	}
+
+	static final class PrefetchConditionalSubscriber<T>
```

Comment:
I have stopped reviewing here for now. Is there any way some code can be shared between the two subscriber classes?